### ISSUE TITLE
Boolean indexing accesses out of bound elements

### ISSUE THREAD

**Issue by leezu** [apache/incubator-mxnet#17990]
## Description
CI with updated toolchain (ie #17984) catches the bug.

`vector: :_M_range_check: __n (which is 2) >= this->size() (which is 2)`

## To Reproduce
Build with this simple patch

``` diff
diff --git a/src/operator/numpy/np_boolean_mask_assign.cc b/src/operator/numpy/np_boolean_mask_assign.cc
index e01ebb7c6..d5ab00835 100644
--- a/src/operator/numpy/np_boolean_mask_assign.cc
+++ b/src/operator/numpy/np_boolean_mask_assign.cc
@@ -220,7 +220,7 @@ void NumpyBooleanAssignForwardCPU(const nnvm::NodeAttrs& attrs,
   // If there's no True in mask, return directly
   if (valid_num == 0) return;

-  const TShape& vshape = inputs[2].shape_;
+  const TShape& vshape = inputs.at(2).shape_;

   if (inputs.size() == 3U) {
     // tensor case
```

OR follow the instructions in https://github.com/apache/incubator-mxnet/issues/17987 to trigger this via glibc assertions in debug build.

Run `test_np_ndarray_boolean_indexing`, `test_npx_batch_dot` or `test_npi_boolean_assign` to trigger the bug.

CC: @haojin2

**Comment by leezu** [comment]
@haojin2 will you fix the bug and reenable the tests?

**Comment by haojin2** [comment]
Fixed in #17796, test re-enabled

**Comment by yzhliu** [comment]
should be fixed by #18158
